File: BOJ/11727/A.py
import sys

sys.stdin = open("D:\Projects\AA\BOJ\\11727\input.txt")
input = sys.stdin.readline

N = int(input().rstrip())
dp = [0] * (N+1)

dp[1] = 1

if N == 1:
    print(dp[1])
    sys.exit(0)

dp[2] = 3
for i in range(3, len(dp)):
    dp[i] = dp[i-1] + dp[i-2] * 2

print(dp[N] % 10007)

File: BOJ/11727/A.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<long long> dp(1001, 0);

int main()
{
    freopen("input.txt", "rt", stdin);
    int n;
    cin >> n;
    dp[1] = 1;
    dp[2] = 3;
    for (int i = 3; i <= n; i++)
    {
        // odd
        if (i % 2)
        {
            dp[i] = (dp[i - 1] * 2 - 1) % 10007;
        }
        // even
        else
        {
            dp[i] = (dp[i - 1] * 2 + 1) % 10007;
        }
    }
    cout << dp[n];
}